pylint/checkers/format.py

```diff
@@ -725,20 +725,49 @@ def check_lines(self, lines: str, lineno: int) -> None:
             - no trailing whitespace
             - less than a maximum number of characters
         """
-        # By default, check the line length
-        check_l_length = True
+        # we're first going to do a rough check whether any lines in this set
+        # go over the line limit. If none of them do, then we don't need to
+        # parse out the pylint options later on and can just assume that these
+        # lines are clean
+
+        # we'll also handle the line ending check here to avoid double-iteration
+        # unless the line lengths are suspect
+
+        max_chars = self.config.max_line_length
+
+        potential_line_length_warning = False
+
+        # hold onto the initial lineno for later
+        starting_lineno = lineno
+        for line in self.specific_splitlines(lines):
+            self.check_line_ending(line, lineno)
+            # this check is purposefully simple and doesn't rstrip
+            # since this is running on every line you're checking it's
+            # advantageous to avoid doing a lot of work
+            if len(line) > max_chars:
+                potential_line_length_warning = True
```

Comment:
Could we `break` here instead  ? Or maybe we could create a list with the potential line length warning index to iterate on them alone later ?

Reply:
Just as a prelude: the "`check_lines` "lines" here are (to my understanding) grouped by fragments (either a line for expressions or the entire docstring for docstrings. But it's not, like, all the lines in a file

Basically I was trying to write this to minimize iterations, and we're doing the line ending check in the same block here. Outright `break`ing would skip over the line ending checks. 

Having a temporary list is a bit tricky too because the "full" check requires stripping of the pylint directives, so indices might shift around. 

I'm already a bit uncomfortable with this code as it's getting a bit "performance over clarity", and I couldn't figure out ways to write it to be more performant without adding a bunch of intermediate concepts. At least from my sort of performance philosophy of "most codebases are passing", I think having the cases where lines _are_ failing (or close to) be slower is OK.

(Warning: the following is annecdotal and I didn't measure it, but I think it's true)
Since this grouping tends to cover only a multiple lines at a time, the lines that _do_ go over the limit will be paired with only a couple other passing lines, and so you won't pay that much of a cost here I think.

If you still think we should try a bit harder to simplify the work here I can look at splitting up the check here a bit more, though. Please tell me if you think it's needed.

Reply:
I'm ok with this clear improvment. I just wonder if we could get rid of `lineno` manipulations through something like this:
```
        for offset, line in numerate(self.specific_splitlines(lines)):
            self.check_line_ending(line, lineno + offset)
            # this check is purposefully simple and doesn't rstrip
            # since this is running on every line you're checking it's
            # advantageous to avoid doing a lot of work
            if len(line) > max_chars:
                potential_line_length_warning = True
```

Reply:
The way I see it, if only one line makes `potential_line_length_warning` be True, then we're going to make the slow regex calculation in any case, so we might as well break the fast for loop and do the slow calculation? Am I reading this wrong @rtpg  ?

Reply:
So the "fast" for loop has a couple parts:
- for each line, check the line ending
- for each line, do the "fast" line length check
- *build the list of lines with self.specific_splitlines`*

I think if we were optimizing for code clarity we would split this into two `for` loops, but I was trying to avoid building up the list of lines several times.

Thinking about this a bit more, and how Python works in general (`self.specific_splitlines` isn't a generator but just returns a list, so it's going to be stored anyways.....) I think we can iterate twice and pay a _bit_ of a cost by iterating twice to get the early bailing behaviour you're asking about.

Reply:
Let me know if I'm wrong but reading what you're saying, a little optimization we could do would be to use a generator for `self.specific_splitlines`, we keep the current line if we break after a fast check and then we do the slow check on the current line + the remaining line of the generator ?

Reply:
yeah we could do something like that. My guess is that it would be nicer for memory usage. This is with a huge asterisk though. Python isn't Rust, it's not going through a bunch of LLVM passes, and it might be a bit of a wash overall (when messing around with some other thing I tried making some generators to cache stuff and mem usage went through the roof because it was written with native Python code). 

Basically gut feeling is that _theoretically_ a generator plus some stuff could be more performant, but code clarity would suffer. And on top of that it might not even be more performant in the end.

Personally I think we can keep the list builder here and iterate twice. We're getting a lot out of this patch already, and we're already at sub-1% part of the runtime so we're hitting diminishing returns for now. I think that for this codebase, and this set of changes, clarity is more important here. And we can come back to this later if we need to.